Unable to install ClickHouse in pfsense FreeBSD

**Describe the unexpected behaviour**
Trying to install clickhouse in FreeBSD (pfsense firewall), following this [link ](https://clickhouse.com/#quick-start)

**How to reproduce**
1 - copy the ClickHouse file into pfsense from [here](https://builds.clickhouse.com/master/freebsd/clickhouse)
2 - chmod a+x ./clickhouse
3 - sudo ./clickhouse install

 **Which ClickHouse server version to use**
- 21.12.1.8928 (official build


**Error message and/or stacktrace**
Code: 107. DB::Exception: Cannot obtain path to the binary from /proc/self/exe, file doesn't exist. (FILE_DOESNT_EXIST) (version 21.12.1.8928 (official build))
![image](https://user-images.githubusercontent.com/14872281/148040641-26f23fc0-39e9-4dcd-97dc-860c9cdefb9b.png)

```./clickhouse server``` command is working, and click-house client is able to connect to server with ```./clickhouse client``` command, 

and the next step is i am running a script (ntopng and click house combination), it is saying 

```
[ERROR] Missing executable clickhouse-client. Install at https://clickhouse.com/#quick-start
```

here is the script link:
https://raw.githubusercontent.com/ntop/ntopng/dev/tools/nindex_export_to_ch.sh

also here is what I am trying to achieve: ntopng with flow dump using ClickHouse in pfsense, if we make this work, this will be a new workaround.

https://www.ntop.org/guides/ntopng/clickhouse.html#migration-from-nindex

--- a/programs/install/Install.cpp
+++ b/programs/install/Install.cpp
@@ -233,6 +233,9 @@ int mainEntryClickHouseInstall(int argc, char ** argv)
             path.pop_back();
 
         fs::path binary_self_path(path);
+#elif defined(OS_FREEBSD)
+        /// https://stackoverflow.com/questions/1023306/finding-current-executables-path-without-proc-self-exe
+        fs::path binary_self_path = "/proc/curproc/file";
 #else
         fs::path binary_self_path = "/proc/self/exe";
 #endif

--- a/programs/install/Install.cpp
+++ b/programs/install/Install.cpp
@@ -191,6 +191,20 @@ static void createUser(const String & user_name, [[maybe_unused]] const String &
 }
 
 
+static std::string formatWithSudo(std::string command, bool needed = true)
+{
+    if (!needed)
+        return command;
+
+#if defined(OS_FREEBSD)
+    /// FreeBSD does not have 'sudo' installed.
+    return fmt::format("su -m root -c '{}'", command);
+#else
+    return fmt::format("sudo {}", command);
+#endif
+}
+
+
 int mainEntryClickHouseInstall(int argc, char ** argv)
 {
     try
@@ -213,10 +227,7 @@ int mainEntryClickHouseInstall(int argc, char ** argv)
 
         if (options.count("help"))
         {
-            std::cout << "Usage: "
-                << (getuid() == 0 ? "" : "sudo ")
-                << argv[0]
-                << " install [options]\n";
+            std::cout << "Usage: " << formatWithSudo(std::string(argv[0]) + " install [options]", getuid() != 0) << '\n';
             std::cout << desc << '\n';
             return 1;
         }
@@ -323,7 +334,7 @@ int mainEntryClickHouseInstall(int argc, char ** argv)
             catch (const Exception & e)
             {
                 if (e.code() == ErrorCodes::CANNOT_OPEN_FILE && geteuid() != 0)
-                    std::cerr << "Install must be run as root: sudo ./clickhouse install\n";
+                    std::cerr << "Install must be run as root: " << formatWithSudo("./clickhouse install") << '\n';
                 throw;
             }
 
@@ -833,9 +844,10 @@ int mainEntryClickHouseInstall(int argc, char ** argv)
             fmt::print(
                 "\nClickHouse has been successfully installed.\n"
                 "\nRestart clickhouse-server with:\n"
-                " sudo clickhouse restart\n"
+                " {}\n"
                 "\nStart clickhouse-client with:\n"
                 " clickhouse-client{}\n\n",
+                formatWithSudo("clickhouse restart"),
                 maybe_password);
         }
         else
@@ -843,9 +855,10 @@ int mainEntryClickHouseInstall(int argc, char ** argv)
             fmt::print(
                 "\nClickHouse has been successfully installed.\n"
                 "\nStart clickhouse-server with:\n"
-                " sudo clickhouse start\n"
+                " {}\n"
                 "\nStart clickhouse-client with:\n"
                 " clickhouse-client{}\n\n",
+                formatWithSudo("clickhouse start"),
                 maybe_password);
         }
     }
@@ -854,7 +867,7 @@ int mainEntryClickHouseInstall(int argc, char ** argv)
         std::cerr << getCurrentExceptionMessage(false) << '\n';
 
         if (getuid() != 0)
-            std::cerr << "\nRun with sudo.\n";
+            std::cerr << "\nRun with " << formatWithSudo("...") << "\n";
 
         return getCurrentExceptionCode();
     }
@@ -910,6 +923,9 @@ namespace
 
         if (!user.empty())
         {
+#if defined(OS_FREEBSD)
+            command = fmt::format("su -m '{}' -c '{}', user, command);
+#else
             bool may_need_sudo = geteuid() != 0;
             if (may_need_sudo)
             {
@@ -920,12 +936,9 @@ namespace
             }
             else
             {
-#if defined(OS_FREEBSD)
-                command = fmt::format("su -m '{}' -c '{}', user, command);
-#else
                 command = fmt::format("su -s /bin/sh '{}' -c '{}'", user, command);
-#endif
             }
+#endif
         }
 
         fmt::print("Will run {}\n", command);
@@ -1129,10 +1142,7 @@ int mainEntryClickHouseStart(int argc, char ** argv)
 
         if (options.count("help"))
         {
-            std::cout << "Usage: "
-                << (getuid() == 0 ? "" : "sudo ")
-                << argv[0]
-                << " start\n";
+            std::cout << "Usage: " << formatWithSudo(std::string(argv[0]) + " start", getuid() != 0) << '\n';
             return 1;
         }
 
@@ -1170,10 +1180,7 @@ int mainEntryClickHouseStop(int argc, char ** argv)
 
         if (options.count("help"))
         {
-            std::cout << "Usage: "
-                << (getuid() == 0 ? "" : "sudo ")
-                << argv[0]
-                << " stop\n";
+            std::cout << "Usage: " << formatWithSudo(std::string(argv[0]) + " stop", getuid() != 0) << '\n';
             return 1;
         }
 
@@ -1206,10 +1213,7 @@ int mainEntryClickHouseStatus(int argc, char ** argv)
 
         if (options.count("help"))
         {
-            std::cout << "Usage: "
-                << (getuid() == 0 ? "" : "sudo ")
-                << argv[0]
-                << " status\n";
+            std::cout << "Usage: " << formatWithSudo(std::string(argv[0]) + " status", getuid() != 0) << '\n';
             return 1;
         }
 
@@ -1248,10 +1252,7 @@ int mainEntryClickHouseRestart(int argc, char ** argv)
 
         if (options.count("help"))
         {
-            std::cout << "Usage: "
-                << (getuid() == 0 ? "" : "sudo ")
-                << argv[0]
-                << " restart\n";
+            std::cout << "Usage: " << formatWithSudo(std::string(argv[0]) + " restart", getuid() != 0) << '\n';
             return 1;
         }
 

--- a/programs/install/Install.cpp
+++ b/programs/install/Install.cpp
@@ -153,10 +153,12 @@ static void createGroup(const String & group_name)
     if (!group_name.empty())
     {
 #if defined(OS_DARWIN)
-
         // TODO: implement.
-
         throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Unable to create a group in macOS");
+#elif defined(OS_FREEBSD)
+        std::string command = fmt::format("pw groupadd {}", group_name);
+        fmt::print(" {}\n", command);
+        executeScript(command);
 #else
         std::string command = fmt::format("groupadd -r {}", group_name);
         fmt::print(" {}\n", command);
@@ -170,10 +172,14 @@ static void createUser(const String & user_name, [[maybe_unused]] const String &
     if (!user_name.empty())
     {
 #if defined(OS_DARWIN)
-
         // TODO: implement.
-
         throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Unable to create a user in macOS");
+#elif defined(OS_FREEBSD)
+        std::string command = group_name.empty()
+            ? fmt::format("pw useradd -s /bin/false -d /nonexistent -n {}", user_name)
+            : fmt::format("pw useradd -s /bin/false -d /nonexistent -g {} -n {}", group_name, user_name);
+        fmt::print(" {}\n", command);
+        executeScript(command);
 #else
         std::string command = group_name.empty()
             ? fmt::format("useradd -r --shell /bin/false --home-dir /nonexistent --user-group {}", user_name)

--- a/programs/install/Install.cpp
+++ b/programs/install/Install.cpp
@@ -235,7 +235,7 @@ int mainEntryClickHouseInstall(int argc, char ** argv)
         fs::path binary_self_path(path);
 #elif defined(OS_FREEBSD)
         /// https://stackoverflow.com/questions/1023306/finding-current-executables-path-without-proc-self-exe
-        fs::path binary_self_path = "/proc/curproc/file";
+        fs::path binary_self_path = argc >= 1 ? argv[0] : "/proc/curproc/file";
 #else
         fs::path binary_self_path = "/proc/self/exe";
 #endif

--- a/programs/install/Install.cpp
+++ b/programs/install/Install.cpp
@@ -924,7 +924,7 @@ namespace
         if (!user.empty())
         {
 #if defined(OS_FREEBSD)
-            command = fmt::format("su -m '{}' -c '{}', user, command);
+            command = fmt::format("su -m '{}' -c '{}'", user, command);
 #else
             bool may_need_sudo = geteuid() != 0;
             if (may_need_sudo)

--- a/programs/install/Install.cpp
+++ b/programs/install/Install.cpp
@@ -919,7 +919,13 @@ namespace
                     command = fmt::format("sudo -u '{}' {}", user, command);
             }
             else
+            {
+#if defined(OS_FREEBSD)
+                command = fmt::format("su -m '{}' -c '{}', user, command);
+#else
                 command = fmt::format("su -s /bin/sh '{}' -c '{}'", user, command);
+#endif
+            }
         }
 
         fmt::print("Will run {}\n", command);

